HTTP & WebSockets request senders.

eugene-babichenko opened the pull request.

eugene-babichenko committed:
STRM2-16 GLOB_RECURSE for project source files

This makes CMakeLists.txt much more compact and readable.

eugene-babichenko committed:
STRM2-16 Use conan to download dependencies

Conan provides standardized approach to use manage dependencies and
using it is faster than downloading dependencies with git.

eugene-babichenko committed:
STRM2-15 Fix builds with macOS GCC and clang

* On macOS it is required to explicitly enable 'safe' functions by
  turning on the __STDC_WANT_LIB_EXT1__ flag.
* `static_assert` must depend on the actual value provided to a
  template. If a condition can be computed at the compile time and is
  known to always fail, compilation will fail even if the invalid
  expression is never constructed.
* Some implicit conversions are not supported by macOS GCC and clang.

Tested with the following compilers:

* Built-in GCC & clang: Apple LLVM version 10.0.1 (clang-1001.0.46.4)
* Homebrew GCC 7.4.0
* Homebrew GCC 9.1.0

eugene-babichenko committed:
cmake: fix CMP0068 warning by using OLD policy

eugene-babichenko committed:
Plain HTTP requests implementation

eugene-babichenko committed:
Implement HTTPS connections.

eugene-babichenko committed:
Cleaner CMake-Conan integration

Switch to using CMAKE_TARGETS argument for conan_cmake_run. This allows
to load header files and link against libraries by listing
Conan-generated targets in target_link_libraries. Using
target_include_directories and find_package is no longer required for
Conan packages.

eugene-babichenko committed:
Convenience wrapper with a struct to perform HTTP requests

eugene-babichenko committed:
Implement shared context for HTTP requests.

eugene-babichenko committed:
Separate headers for HTTP implementation.

eugene-babichenko committed:
Implement WebSockets connection

eugene-babichenko committed:
STRM2-20 Implement getBlockchainHeight

In addition to the method implementation itself, this change set also
contains the following:

* Example targets (turned off by default and can be built by setting
  NEM2_SDK_BUILD_EXAMPLE to ON).
* The basic client structure, similar to Go SDK.
* performHTTPRequest now throws an exception for invalid response
  statuses.

eugene-babichenko committed:
STRM2-20 Implement getBlockByNumber

In addition, this change set contains block data objects and DTOs
implementation.

eugene-babichenko committed:
STRM2-20 Persist context in the client instead re-creating it

eugene-babichenko committed:
STRM2-20 Introduce RequestParamsBuilder for cleaner client code.

eugene-babichenko committed:
STRM2-20 Implement getBlocksByHeightWithLimit

Additionally implemented items:

* MultipleBlocksDto for parsing JSON with multiple blocks at once.
* Support for HTTP redirects.

pananton commented on CMakeLists.txt:
```diff
@@ -108,12 +114,12 @@ target_include_directories(nemcpp
 
-# adding alias for target (required to consistent use of SDK when
-# it is built as subproject)
-add_library("NEM2::nemcpp" ALIAS nemcpp)
```
Probably you accidently removed this line - it allows in-source sdk build. Better to add it back.

eugene-babichenko replied:
Thanks, really accidentally.

pananton reviewed (commented):
@eugene-babichenko I see that you use boost library for json parsing. I suggest you to use sdk classes to parse json. They will be used for transaction parsing and all json endpoints in NEM sdk. Better to have similiar implementation. Let me explain shortly how it was intended to be implemented using your type BlockMeta as example:

* There is a template class VariadicStruct that allows you to describe arbitraty json objects (look into infrastructure/json/parser.h in develop branch head)

* Create DTO for BlockMeta in the following way: 
```
using BlockMetaDTO = VariadicStruct<
    Field<STR_LITERAL("hash"), std::string>,
    Field<STR_LITERAL("generationHash", std::string>,
    Field<STR_LITERAL("totalFee"), Uint64>,
    Field<STR_LITERAL("numTransactions"), Uint64>>;
```

* Use json::Parser Read/Write methods with this type:
```
BlockMetaDTO dto;
Parser::Read(dto, jsonStr);
```

* You can access DTO field values using 'value' method and check whether it were set in json using 'isSet' method. Then you can convert DTO to your user-firendly type BlockMeta.

```
BlockMeta meta = {
    dto.value<"hash"_>(),
    dto.value<"generationHash"_>(),
    dto.value<"totalFee"_>(), // note how Uint64 type from dto ([high, low] in json automatically converted to uint64_t
    dto.value<"numTransactions"_>()
};
```

If you will use this highly recommended approach, declare your DTOs in infrastructure/json/dto folder. If you have any isuues on json::Parser (bugs, use-cases) - contact me. I've tested it a lot, but still could miss something. Better to merge current develop branch to you to get last json::Parser modificatiaons.

eugene-babichenko commented:
@pananton13 Thanks for the explanation/suggestion.

The guide for using those classes as well as the documentation for them is really missing. Can you take a little time for adding them right into the library? That would be really helpful for further development, because the serialization code looks a bit complicated.

pananton commented:
I'll try to do this soon. Currently it has a comment about what types are supported for VariadicStruct fields to be used in serialization in json/Parser.h . To have more idea about how it is used look into already implemented binary transaction DTOs in infrastructure/binary/dto/transaction_dto.h - binary serialization is build on same principles. In general, you describe structure of your json object in terms of VariadicStruct (which is similar to C++ struct, but each field also contains a descriptor, which can hold additional information about it (like whether field is optional in json or not)) and json::Parser can read/write any properly described struct.

eugene-babichenko committed:
Update build documentation.

vulfke482 committed:
Added custom json serialization and translated messages

vulfke482 committed:
Updated variadic_struct.h from develop

vulfke482 committed:
Remade serialization via custom serialization method

vulfke482 committed:
backup

vulfke482 committed:
second try to save commit while uint64 doesn't work

pananton commented on include/nemcpp/message.h:
```diff
@@ -0,0 +1,181 @@
+
+#pragma once
+
+#ifndef NEM2_SDK_CPP_MESSAGE_H
```
#pragma once is used through all sdk headers, not necessary to use include guards

vulfke482 replied:
Thanks, removed.

pananton commented on src/infrastructure/dto/block_dto.h:
```diff
@@ -3,23 +3,56 @@
+using Uint64 = nem2_sdk::internal::json::Uint64;
+
 
 namespace nem2_sdk::internal::dto {
```
Use internal::json namespace for json DTOs to stay consistent with binary DTOs which are placed in internal::binary namespace.

vulfke482 replied:
Thanks for notice, in future I will resolve it.

pananton replied:
One more note) Sorry for being annoying - after all it's up to you to decide how you will continue develop proxima cpp sdk - but I've noticed that you did not follow existing naming style. In other SDKs (java, ts) types that are used for serialization/deserialization are called SomeTypeDTO, while corresponding user-friendly types that are exported by SDK simply SomeType. You name DTOs like SomeTypeDtoT and exported types (if I'm not mistaken) SomeTypeDto. Better if all SDKs will be as much similar as possible)

vulfke482 replied:
No, SomeTypeDto is a wrapper of SomeType with functions for processing json. It was made in order to separate json processing from Some Type objects. And DTOT's are used as a tool for extracting data from json. The need of wrapper class could be controversial, but it is quite conveniently. I agree that Dto's functions must return SomeType and I will correct them. Thanks, for notice)

vulfke482 replied:
In future it really could be simplified via one template function from_json, and several getSomeType or get<SomeType>. I will think about it when i have some time)

pananton replied:
Ok)

pananton commented on src/infrastructure/utils/variadic_struct.h:
```diff
@@ -364,9 +367,9 @@ namespace nem2_sdk { namespace internal {
 	
 	template<size_t I> struct struct_field_by_index<I, NullStruct> {
-		static_assert(I < 0, "variadic struct field index is out of range");
+		static_assert(sizeof(I) == 0, "variadic struct field index is out of range");
```
There can be a problem with this line when building on mac. @eugene-babichenko fixed it in his branch, you should change condition from ```sizeof(I) == 0``` to ```I < 0```

vulfke482 replied:
Done)

pananton commented:
@vulfke482 Oops, looks like you've accidently pushed all build artifacts to repo. They should be removed before your branch get merged to develop. Note, that if you simple remove files from your local copy and then push - it will not solve the problem, because binary files will stay in repo history, increasing the size of repo. Look for instructions on git filter-branch command here: https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History .

vulfke482 committed:
It is a half a way for masaic, it is compilling anyway

vulfke482 committed:
Ended skeleton for mosaics

vulfke482 committed:
All review notices, except dto architecture and build artefacts

vulfke482 committed:
Mosaics are finished

vulfke482 committed:
Resolved problems with DTO

xgreenx commented on CMakeLists.txt:
```diff
@@ -185,7 +92,13 @@ if (NEM2_SDK_BUILD_STATIC)
 		OUTPUT_NAME "nemcpp-static")
 else()
-	add_library(nemcpp SHARED "")
+<<<<<<< HEAD
```
Something is wrong here

xgreenx commented on CMakeLists.txt:
```diff
@@ -265,3 +178,18 @@ if (NEM2_SDK_BUILD_TESTS)
 endif()
+
+if (NEM2_SDK_BUILD_EXAMPLES)
+<<<<<<< HEAD
```
And here

xgreenx commented on CMakeLists.txt:
```diff
@@ -265,3 +178,18 @@ if (NEM2_SDK_BUILD_TESTS)
+	add_executable(get_block ${CMAKE_CURRENT_SOURCE_DIR}/examples/get_block.cpp include/nemcpp/message.cpp include/nemcpp/message.h include/nemcpp/client/mosaic_service.h)
+
+=======
+	add_executable(get_block ${CMAKE_CURRENT_SOURCE_DIR}/examples/get_block.cpp include/nemcpp/message.cpp include/nemcpp/message.h)
```
And here

xgreenx commented on examples/get_block.cpp:
```diff
@@ -0,0 +1,23 @@
+
+	auto blocks = client->blockchain()->getBlocksByHeightWithLimit(height - 1, 25);
+	for (auto& block: blocks) {
+        std::cout << "Block signature: " << block.block.signature << std::endl;
```
indent

xgreenx commented on examples/get_block.cpp:
```diff
@@ -0,0 +1,23 @@
+
+	auto height = client->blockchain()->getBlockchainHeight();
+	std::cout << "Block number: " << height << std::endl;
+//    height = 148983;
```
I think you need to remove commented code

xgreenx commented on include/nemcpp/client.h:
```diff
@@ -0,0 +1,33 @@
+#include <nemcpp/config.h>
+#include <nemcpp/client/blockchain_service.h>
+
+#include <memory>
```
3x memory include

xgreenx commented on include/nemcpp/client/mosaic_service.h:
```diff
@@ -0,0 +1,19 @@
+//
+// Created by vulfke on 23.07.19.
```
/**
*** Copyright 2019 ProximaX Limited. All rights reserved.
*** Use of this source code is governed by the Apache 2.0
*** license that can be found in the LICENSE file.
**/

xgreenx commented on include/nemcpp/model/mosaic/mosaic.h:
```diff
@@ -12,20 +12,53 @@ namespace nem2_sdk {
+
+    };
+
+    class RichMosaic {
```
I think will be better to use name from the swagger like MosaicInfo

xgreenx commented on include/nemcpp/model/mosaic/mosaic.h:
```diff
@@ -12,20 +12,53 @@ namespace nem2_sdk {
+        std::string name;
+    };
+
+	class MosaicMeta {
```
Indent

xgreenx commented on include/nemcpp/config.h:
```diff
@@ -41,6 +42,11 @@ namespace nem2_sdk {
 		/// Determines transaction deadline which is used by default.
 		NetworkDuration TransactionDeadline = Transaction_Deadline;
+
+		std::string nodeAddress = "localhost";
```
Why do you can to have only one config?
You can easy have many instance of different config related to different networks.

pananton replied:
To be done.

pananton replied:
Also please take a look, how configuration parameters and transactions are named. Having one style on  line 44 and another on 46 is strange.

xgreenx commented on include/nemcpp/message.h:
```diff
@@ -0,0 +1,180 @@
+        }
+
+        std::string String() const override {
+            std::ostringstream result;
```
I think you need to move it to utils

pananton replied:
If I'm not mistaken, this type is inteded to be used for representing message in transfer transaction. In that case better to move it to model/transaciton (it will no be used anywhere except transfer tx). Also note,  that other sdk types (at least - typescript, have not looked into other) place same type there. Of course, it definetly should not be under root.

xgreenx commented on include/nemcpp/message.h:
```diff
@@ -0,0 +1,180 @@
+        }
+
+        std::string Message() {
+            std::ostringstream result;
```
I'm not sure about Message method, but you can call String() here

xgreenx commented on include/nemcpp/message.h:
```diff
@@ -0,0 +1,180 @@
+    };
+
+
+    class Message {
```
Virtual destructor

xgreenx commented on include/nemcpp/message.h:
```diff
@@ -0,0 +1,180 @@
+        using PlainMessagePtr = std::shared_ptr<PlainMessage>;
+    public:
+        PlainMessage(const std::string& payload_string) {
+            for(auto x : payload_string) {
```
https://ru.cppreference.com/w/cpp/container/vector/vector

xgreenx commented on include/nemcpp/message.h:
```diff
@@ -0,0 +1,180 @@
+
+    class SecureMessage : public Message {
+    public:
+        SecureMessage(const ByteArray& encodedList) {
```
I think you need to std::move encoded data instead of copy

xgreenx commented on include/nemcpp/message.h:
```diff
@@ -0,0 +1,180 @@
+        }
+
+        std::string String() const override {
+            std::ostringstream result;
```
Same

xgreenx commented on include/nemcpp/model/blockchain/block.h:
```diff
@@ -0,0 +1,32 @@
+		std::string beneficiaryPublicKey;
+	};
+
+	class Block {
```
Block.block It is strange=) BlockInfo is better, or maybe you can move all fields from BlockData to Block

xgreenx commented on include/nemcpp/model/transaction/transfer_transaction.h:
```diff
@@ -79,7 +79,7 @@ namespace nem2_sdk {
 	/// \note Optional transaction parameters are initialized using \c Config if not set explicitly.
 	std::unique_ptr<EmbeddedTransferTransaction>
 	CreateEmbeddedTransferTransaction(const Address& recipient,
```
I'm not sure that users of SDK need to know about EmbeddedTransaction.

pananton replied:
What type will have embedded transactions then? Same as not embedded, i.e. instead of EmbeddedTransferTransaction users will get from AggregateTransaction container of TransferTransaction ? Then plenty of methods, inherited from base class for general transactions will have no meaning at all: fee, deadline, signature, hash, sign/verify etc. Maybe I'm missing some elegant solution for this though (maybe you have one?). And I know, that this approach is used in ts sdk and java sdk, but I found it not ideal too. Users of this sdks still need to know, that TransferTransaction they get from AggregateTransaction can't be treated totally the same as "simple" TransferTransaction.
The drawback of current solution is that it's hard to treat embedded and general transactions in the same way without template magic, because they can't be converted between each other (same inheritance mechanism is used for them as in catapult). I had and alternative (but is uses multiple inheritance) and will probably implement it in nem sdk repository later.

vulfke482 committed:
Added methods for chain score and storage info requests.

vulfke482 committed:
Changes in masaics and mesasges after review

vulfke482 committed:
Debuging problems with server

vulfke482 committed:
Updates:
Implemented namespace api infrastructure.
Implemented account api infrastructure.
Resolved noticed in review problems.
Deleted lower_higher.h/.cpp
Added exception processing.

vulfke482 committed:
Minor changes in mosaic service

vulfke482 committed:
Changed input parameters in masaic api requests.

vulfke482 committed:
Added virtual destructor to messages.h:Message.

vulfke482 committed:
Namespace queries are done. Beginning of refactoring.

vulfke482 committed:
Refactoring is complited. Cannot link 'lib/libnemcpp.so.1.0.0'

vulfke482 committed:
Refactoring is done.

vulfke482 committed:
Added account and network services

vulfke482 committed:
Added general transaction dto.

vulfke482 committed:
Transactions: draft

vulfke482 committed:
backup

eugene-babichenko committed:
fix build: JSON parser and private key

eugene-babichenko committed:
Merge pull request #9 from proximax-storage/feature/update-build-docs

Build docs update

eugene-babichenko committed:
Merge branch 'develop' into feature/clang-build

eugene-babichenko committed:
Merge pull request #1 from proximax-storage/feature/clang-build

Fix builds for macOS

vulfke482 committed:
backup2

eugene-babichenko committed:
Fix builds with GCC

Move namespace abbreviations so they don't conflict with newer ones in
cpp files. Allow to build without WebSockets (they are disabled by
default).

vulfke482 committed:
Trying to resolve problems with binary serialization

vulfke482 committed:
before deleting all binary serialization

vulfke482 committed:
Almost builded

vulfke482 committed:
Almost build 2

vulfke482 committed:
Working version

vulfke482 committed:
GET and POST requests are working

vulfke482 committed:
Added optional fields in DTO

eugene-babichenko committed:
fix build on macOS

eugene-babichenko committed:
Merge pull request #13 from proximax-storage/feature/http-websockets-impl

Implement HTTP connection & add a sample request implementation

vulfke482 committed:
Added several overloaded methods

vulfke482 committed:
Rebase is done, but there are some problems with connection.

vulfke482 committed:
Removed cmake-build-debug
